### PR TITLE
A Compressed Whisper Format: cwhisper

### PR DESCRIPTION
**Pull request by bom-d-van**
Hi maintainers,

TLDR: This is a long and big PR aiming at bringing compression to whisper (thus go-carbon) while being backward compatible, also including some necessary tooling for debugging and migration.

Compression algorithm source: 4.1 Time series compression in [Gorilla: A Fast, Scalable, In-Memory Time Series Database](https://www.vldb.org/pvldb/vol8/p1816-teller.pdf)

The compressed format is dubbed as cwhisper.

### How does it work in a nutshell

An example format: https://github.com/bom-d-van/go-whisper/blob/master/doc/compressed.md

In cwhisper, archives are broken down into multiple blocks (by default 7200 data points per block as recommended by the gorilla paper), and data points are compressed into blocks. cwhisper assumes 2 as the default data point size, but when it detects that the default size is too small, it would grow the file.

cwhisper still has one file per metric, it's doing round-robin update, instead of rotating data points, block is rotation unit for archives.

#### Compression ratio

Data point in cwhisper ranges from 2 - 14 bytes (12 bytes for standard format). So in theory, cwhisper file size could be 16.67% - 116.67% of standard file size. So the theoretical compression ratio is 6 - 0.86.

In random data point testing, compressed/uncompressed ratio is between 18.88% and 113.25%. 

In real production payload, we are seeing 50%+ less disk space usage.

#### Read/Write Performance

Some simple benchmarks are made for comparison, cwhisper is actually faster than standard whisper format in both read/write (with tradeoff of course):

```
BenchmarkWriteStandard-8           	   50000	     33824 ns/op
BenchmarkWriteCompressed-8         	 1000000	      1630 ns/op

BenchmarkReadStandard-8            	     500	   2270392 ns/op
BenchmarkReadCompressed-8          	   10000	    260862 ns/op
```

#### Drawbacks

* cwhisper is faster and smaller, but unlike standard format, you can't easily backfill/update/rewrite old data points because it's not data-point addressable.
* file size could grow if data points are irregular.

#### Suitable Application

cwhisper is most suitable for metrics that are mostly regular and less likely needed to backfill/rewrite old data, like system metrics. cwhisper also works nicely for sparse metrics.

#### Pull Request Breakdown

| Files | Description |
| --- | --- |
| compress.go, compress_test.go | the core changes are done in these files |
|whisper.go|major API calls and Options are modified to support compressed format|
|cmd/*.go, debug.go, debug_test.go|handy tools for migration and debugging|
|internal|some work-in-progress fuzz logics|

### CODE REVIEW

File: whisper.go [hunk 1]
```diff
@@ -127,22 +150,47 @@ func ParseRetentionDefs(retentionDefs string) (Retentions, error) {
...
+
+	opts     *Options
+	Extended bool
 }
 
 // Wrappers for whisper.file operations
 func (whisper *Whisper) fileWriteAt(b []byte, off int64) error {
 	_, err := whisper.file.WriteAt(b, off)
+	if err != nil {
+		panic(err)
```
Comment (Civil) [inline whisper.go]: What's the point to panic here and then return 'err'?

For libraries it's usually better to return error if any.

Comment (bom-d-van) [reply whisper.go]: Thanks for pointing it out, that's just debug code. Removed.

**Comment by bom-d-van** [comment]
Hi, I have marked this PR as ready to review. Could you take a look and let me know if there is any concerns? 
Meanwhile, I will make another PR soon to github.com/lomik/go-carbon so we know how the integration looks like.

**Comment by Civil** [comment]
Well, I've only checked so far that it doesn't do any incompatible changes. As the PR is quite huge I think we have two options:
1. I can just merge it as it is, and open issue if I'll find anything weird in the new code.
2. Or we can wait until I go through all of it... I will try to do it fast, but I can't give any guarantees.

Only concern I have so far is that if you are changing the format of the archive anyway, it may be beneficial to include some sort of archive version in new metadata for the case you want to change it again in the future. But that's just a suggestion.

P.s. anyway, I've sent you an invite to give committer rights to this repo as I think it would be better if you'll continue maintaining the code once it's done.

**Comment by bom-d-van** [comment]
Hi @Civil , thanks for your reply. I understand this is quite a huge PR and has many new details, so I really appreciate your review. We don't have to rush. Both options are good for me. Having extra eyes going through the changes is beneficial. 

How about we give it another week, by that time if we don't see anything blocking, then we proceed to merge it? 

> Only concern I have so far is that if you are changing the format of the archive anyway, it may be beneficial to include some sort of archive version in new metadata for the case you want to change it again in the future. But that's just a suggestion.

There is a version field in the new format which we could used for this request: https://github.com/go-graphite/go-whisper/pull/1/files#diff-a17e829d9b187451ae1b61fb70de3654R178

> P.s. anyway, I've sent you an invite to give committer rights to this repo as I think it would be better if you'll continue maintaining the code once it's done.

Yes, this would be great. And I will continue to use PR for new changes so that you and other maintainers know what's happening.

**Comment by bom-d-van** [comment]
Hi @Civil , thanks for your reply. I understand this is quite a huge PR so I really appreciate your review. We don't have to rush, it's nice to have extra eyes going through the changes. 

How about we give it one more week, and if you don't see blocking things by that time or you don't manage to find time to go through this, then we proceed with a merge. I am happy to maintain this feature, so we can't always open issues if we find any for it later on. 

> Only concern I have so far is that if you are changing the format of the archive anyway, it may be beneficial to include some sort of archive version in new metadata for the case you want to change it again in the future. But that's just a suggestion.

There is a version field in the new format which we could used for this feature.

> P.s. anyway, I've sent you an invite to give committer rights to this repo as I think it would be better if you'll continue maintaining the code once it's done.

I haven't receive the invitation yet, but this is great to me, thanks!

**Comment by bom-d-van** [comment]
Hi @Civil, I will proceed with merging these changes to the library unless you object. As you suggested, we can always open issues when we discover problems.

Also, for all up-coming fixes/changes to the library, I will still go through PR just so that all maintainers know what's going on.

**Comment by dgryski** [comment]
Has the implementation been fuzzed?

**Comment by dgryski** [comment]
A few quick notes:
   - staticcheck finds some issues
   - should the BitWriter type really be exported and part of the public API?
   - A benchmarking framework + numbers would be nice

**Comment by bom-d-van** [comment]
Hi @dgryski , thanks for the notes. :)

> staticcheck finds some issues

I did briefly used staticcheck but I was seeing lots stdlib warnings and I ended up just fix go vet errors. But seems using specific filenames show me something. Will see how many I could fix.

> should the BitWriter type really be exported and part of the public API?

Good point. Mistakes made in early days. Gonna un-export it.

> And some fuzzing logic. Round-trip fuzzing the compressor would also be good.

The fuzzing logics there isn't really working (especially for the whole file, fuzzing just archive helps a bit) but it's something I would like to optimize in the future. Maybe [structural aware fuzzing](https://github.com/google/fuzzer-test-suite/blob/master/tutorial/structure-aware-fuzzing.md) would help?
Round-trip fuzzing is a good idea. Will make a quick one.
Also, there is some tests with [random input](https://github.com/go-graphite/go-whisper/pull/1/files#diff-f1b0dc22562f64cc159c23d50b65f5b1R322) that did catch more problems in my code.

**Comment by bom-d-van** [comment]
Didn't find good staticcheck errors, still seeing only stdlib ones. Maybe something is wrong with my staticcheck. But it's bit{Reader,Writer} now.

**Comment by dgryski** [comment]
```
compress.go:1378:3: this value of err is never used (SA4006)
compress.go:1392:5: this value of err is never used (SA4006)
```

**Comment by bom-d-van** [comment]
@dgryski thank you, that's great. Fix pushed. Also switching to a pre-compile staticcheck and can see stuff now (couldn't stop wondering why go-get-ed staticcheck doesn't show me right stuff; probably dependencies?).

**Comment by bom-d-van** [comment]
@dgryski @Civil I'm gonna request to merge this PR again so that I could continue with another change sets for supporting bucky-tool, in order to have this new format being able to move forward in production.

So if you do not object, I will merge this soon. However you could still comment here or open issues if you think something is bad; I will certainly follow up on them. Really appreciate every feedback I could get.

---

What's more, with my new understanding gained through working on cwhisper, I have new proposal about supporting percentiles in whisper thus in Graphite too. It could be interesting.

**Comment by dgryski** [comment]
If you're running this in production and are happy with it, I'm fine to merge.  I think @Civil gave you a commit bit here?

**Comment by bom-d-van** [comment]
(We do have a box running this feature informally. ;))

Yep, I have the commit bit. Just want to let you guys know because of your helps. 
It's not perfect for sure, but we can always improve it along the way.

Will merge it soon and start new works on branches in this repo. :D

**Comment by Civil** [comment]
I'm ok with merging it.

**Comment by Quentin-M** [comment]
Amazing feature, thanks for all the hard work!

Note that unfortunately we can't use it due to the lack of support for a/ cwhisper in python whisper b/ tags in go carbon / carbonapi (https://github.com/lomik/go-carbon/issues/259). Because of b/, we must use a sketch hybrid solution where go-carbon is used for writes (can be compressed) but graphite-web for reads (can't read compressed).

**Comment by bom-d-van** [comment]
Hi @Quentin-M , thanks for your compliment. We would be thrilled to see more people using it.

There are some level of tagging support in go-carbon, but I'm not sure if it's fully compatible with the python implementation. If it's, maybe you could let graphite-web talk to go-carbon. 
@azhiltsov might have something to share.
